apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Cloud.Spanner.V1.Logging;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+using Grpc.Core;
+
+// ReSharper disable PrivateFieldCanBeConvertedToLocalVariable
+// ReSharper disable NotAccessedField.Local
+// ReSharper disable EmptyDestructor
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// Provides streaming access to a Spanner SQL query that automatically retries, handles
+    /// chunking and recoverable errors.
+    /// </summary>
+    public sealed class ReliableStreamReader : IDisposable
+    {
+        private AsyncServerStreamingCall<PartialResultSet> _currentCall;
+        private readonly object _sync = new object();
+        private readonly SpannerClient _spannerClient;
+        private readonly ExecuteSqlRequest _request;
+        private readonly Session _session;
+        private readonly CallSettings _callTiming;
+        private ResultSetMetadata _metadata;
+        private bool _isReading = true;
+
+        private ByteString _resumeToken;
+        private int _resumeSkipCount;
+
+        internal ReliableStreamReader(SpannerClient spannerClientImpl, ExecuteSqlRequest request, Session session)
```

Comment:
It may be better to not to take the request in the first place and let it be constructed internally.
This will be a good thing to discuss when we're ready to make public like you said.